examples/joke_bot/src/main.rs

```diff
@@ -0,0 +1,270 @@
+use tokio_stream::StreamExt as _;
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn Error>> {
+    if let Err(e) = run().await {
+        eprintln!("{}", e);
+        exit(1)
+    }
+    Ok(())
+}
+
+type HttpClient = client::http_client::HyperNativeTls;
+type MatrixClient = client::Client<client::http_client::HyperNativeTls>;
+
+async fn run() -> Result<(), Box<dyn Error>> {
+    let config =
+        read_config().await.map_err(|e| format!("configuration in ./config is invalid: {}", e))?;
+    let http_client =
+        hyper::Client::builder().build::<_, hyper::Body>(hyper_tls::HttpsConnector::new());
+    let matrix_client = if let Some(state) = read_state().await.ok().flatten() {
+        MatrixClient::with_http_client(
+            http_client.clone(),
+            config.homeserver.to_owned(),
+            Some(state.access_token),
+        )
+    } else if config.password.is_some() {
+        let client = create_matrix_session(http_client.clone(), &config).await?;
+
+        if let Err(err) = write_state(&State {
+            access_token: client.access_token().expect("Matrix access token is missing"),
+        })
+        .await
+        {
+            eprintln!("Failed to persist access token to disk. Re-authentication will be required on the next startup: {}", err)
+        }
+        client
+    } else {
+        return Err("No previous session found and no credentials stored in config".into());
+    };
+
+    let filter = FilterDefinition::ignore_all().into();
+    let initial_sync_response = matrix_client
+        .send_request(assign!(sync_events::Request::new(), {
+            filter: Some(&filter),
+        }))
+        .await?;
+    let user_id = &config.username;
+    let not_senders = &[user_id.clone()];
+    let filter = {
+        let mut filter = FilterDefinition::empty();
+        filter.room.timeline.not_senders = not_senders;
+        filter
+    }
+    .into();
+
+    let mut sync_stream = Box::pin(matrix_client.sync(
+        Some(&filter),
+        initial_sync_response.next_batch,
+        &PresenceState::Online,
+        Some(Duration::from_secs(30)),
+    ));
+    println!("Listening...");
+    while let Some(response) = sync_stream.try_next().await? {
+        println!("{}", response.next_batch);
+        for (room_id, room_info) in response.rooms.join {
+            for e in &room_info.timeline.events {
+                match handle_messages(&http_client, &matrix_client, e, &room_id, user_id).await {
+                    Ok(_) => {}
+                    Err(err) => {
+                        eprintln!("failed to respond to message: {}", err)
+                    }
+                }
+            }
+        }
+
+        for (room_id, _) in response.rooms.invite {
+            match handle_invitations(&http_client, &matrix_client, &room_id).await {
+                Ok(_) => {}
+                Err(err) => eprintln!("failed to accept invitation for room {}: {}", &room_id, err),
+            }
+        }
```

Comment:
This does a lot of waiting sequentially that could be done concurrently. That is, waiting for messages to be sent and invitation / joke API requests to go through. I think it would be nice to use `tokio::spawn` or something like `join_all` / `FuturesUnordered` from `futures-util` for this.

Doesn't have to block this PR though, we can do it later as well.

Reply:
Sounds good. Let's revisit that later.